core/types/src/key_request.rs

```diff
@@ -0,0 +1,273 @@
+    SGX_KEY_REQUEST_RESERVED2_BYTES,
+};
+use rand::{CryptoRng, RngCore};
+
+#[repr(u16)]
+pub enum KeyName {
+    /// Launch key
+    EnclaveInitializationToken = SGX_KEYSELECT_EINITTOKEN,
+
+    /// Provisioning Key
+    Provision = SGX_KEYSELECT_PROVISION,
+
+    /// Provisioning Seal Key
+    ProvisionSeal = SGX_KEYSELECT_PROVISION_SEAL,
+
+    /// Report Key
+    Report = SGX_KEYSELECT_REPORT,
+
+    /// Seal Key
+    Seal = SGX_KEYSELECT_SEAL,
+}
+
+#[repr(transparent)]
+#[derive(Clone, Hash, PartialEq, Eq, Debug, Copy)]
+pub struct KeyPolicy(u16);
+impl KeyPolicy {
+    /// Derive key using the enclave's ENCLAVE measurement register
+    pub const MRENCLAVE: KeyPolicy = KeyPolicy(SGX_KEYPOLICY_MRENCLAVE);
+
+    /// Derive key using the enclave's SINGER measurement register
+    pub const MRSIGNER: KeyPolicy = KeyPolicy(SGX_KEYPOLICY_MRSIGNER);
+
+    /// Derive key without the enclave's ISVPRODID
+    pub const NO_ISV_PROD_ID: KeyPolicy = KeyPolicy(SGX_KEYPOLICY_NOISVPRODID);
+
+    /// Derive key with the enclave's CONFIGID
+    pub const CONFIG_ID: KeyPolicy = KeyPolicy(SGX_KEYPOLICY_CONFIGID);
+
+    /// Derive key with the enclave's ISVFAMILYID
+    pub const ISV_FAMILY_ID: KeyPolicy = KeyPolicy(SGX_KEYPOLICY_ISVFAMILYID);
+
+    /// Derive key with the enclave's ISVEXTPRODID
+    pub const ISV_EXTENDED_PROD_ID: KeyPolicy = KeyPolicy(SGX_KEYPOLICY_ISVEXTPRODID);
+}
+
+impl core::ops::BitOr<KeyPolicy> for KeyPolicy {
+    type Output = Self;
+    #[inline]
+    fn bitor(self, other: Self) -> Self {
+        KeyPolicy(self.0 | other.0)
+    }
+}
+impl ::core::ops::BitOrAssign for KeyPolicy {
+    #[inline]
+    fn bitor_assign(&mut self, rhs: KeyPolicy) {
+        self.0 |= rhs.0;
+    }
+}
+impl ::core::ops::BitAnd<KeyPolicy> for KeyPolicy {
+    type Output = Self;
+    #[inline]
+    fn bitand(self, other: Self) -> Self {
+        KeyPolicy(self.0 & other.0)
+    }
+}
+impl ::core::ops::BitAndAssign for KeyPolicy {
+    #[inline]
+    fn bitand_assign(&mut self, rhs: KeyPolicy) {
+        self.0 &= rhs.0;
+    }
+}
+
+new_type_wrapper! {
+    ConfigSvn, sgx_config_svn_t;
+}
+
+// Suppress clippy as the `new_type_wrapper` macro can't derive Default for many
+// of the types.
+#[allow(clippy::derivable_impls)]
+impl Default for ConfigSvn {
+    fn default() -> Self {
+        ConfigSvn(0)
+    }
+}
+
+new_type_wrapper! {
+    IsvSvn, sgx_isv_svn_t;
+}
+
+#[allow(clippy::derivable_impls)]
+impl Default for IsvSvn {
+    fn default() -> Self {
+        IsvSvn(0)
+    }
+}
+
+new_type_wrapper! {
+    CpuSvn, sgx_cpu_svn_t;
+}
+
+impl Default for CpuSvn {
+    fn default() -> Self {
+        CpuSvn(sgx_cpu_svn_t {
+            svn: [0; SGX_CPUSVN_SIZE],
+        })
+    }
+}
+
+new_type_wrapper! {
+    KeyRequest, sgx_key_request_t;
+}
+
+impl KeyRequest {
+    /// Creates a new [KeyRequest].
+    ///
+    /// The [KeyRequest] will be mostly empty and can be built up using the
+    /// setter methods like [KeyRequest::set_isv_svn]
```

Comment:
I didn't think of it before, but the rust way to do this is to have a `KeyRequestBuilder`, e.g.:

`KeyRequestBuilder::new(&csprng).key_name(foo).isv_svn(bar)`

Then you can nuke all the setters and just leave this as an opaque struct (or add getters)

Reply:
created a builder for this.  Note Attributes is still a setter style, I'm thinking since it has a default of all 0s and with only 2 setters it may serve as is.